hermes/oauth_clients.py

```diff
@@ -0,0 +1,84 @@
+from django.conf import settings
+from datetime import datetime
+import logging
+from django.utils import timezone
+from authlib.integrations.django_client import OAuth
+from hermes.models import OAuthToken
+
+logger = logging.getLogger(__name__)
+
+
+def update_token(user, integrated_app, token_details, group_permissions=[]):
+    """ Called on authentication of an oauth service, this method creates or updates the already
+        created OAuthToken with the details received from the authentication service.
+    """
+    # enforce the restriction of only one stored token per user and integrated app
+    OAuthToken.objects.update_or_create(user=user, integrated_app=integrated_app, defaults={
+        'access_token': token_details.get('access_token'),
+        'refresh_token': token_details.get('refresh_token'),
+        'token_type': token_details.get('token_type'),
+        'group_permissions': group_permissions,
+        'expires_at': datetime.fromtimestamp(token_details.get('expires_at')).replace(tzinfo=timezone.utc),
+        'expires_in': token_details.get('expires_in')
+    })
+
+
+def refresh_token(user, token):
+    """ Takes in a user and token object and attempts to refresh the token. The token object must
+        have a valid refresh_token stored in it, and the oauth_client must support refreshing its
+        access_token. Updates the access_token and expiration times if successful, and returns back
+        the updated token object.
+    """
+    logger.info(f"Refreshing {token.integrated_app} token for user {user.username}")
+    oauth_client = oauth_clients.get(token.integrated_app)
+    if oauth_client:
+        # The refreshed token details contain only an access token and expiration
+        try:
+            token_details = oauth_client.fetch_access_token(
+                refresh_token=token.refresh_token, grant_type='refresh_token'
+            )
+            if not token_details:
+                # If the token_details are not returned, assume the refresh token is bad and delete the entire
+                # OAuthToken
+                # TODO: I don't actually know what is returned if the refresh_token expires so this might not work!
```

Comment:
Should we fix this todo?

Reply:
I still don't actually know this yet... I tried to program defensively here, but the token expiration is 30 days so I still haven't encountered this happening yet.

Reply:
According to the [OpenID Connect specification](https://openid.net/specs/openid-connect-core-1_0.html#RefreshErrorResponse):

> If the Refresh Request is invalid or unauthorized, the Authorization Server returns the Token Error Response as defined in Section 5.2 of [OAuth 2.0](https://openid.net/specs/openid-connect-core-1_0.html#RFC6749) [RFC6749].

And the error response is documented by the RFC here: https://datatracker.ietf.org/doc/html/rfc6749#section-4.1.2.1

Reply:
yeah as far as I can tell authlib will throw an Exception of some type in the case of an error, which we catch and delete the token on, forcing the user to re-authenticate from their profile. Just haven't actually tested it yet.

Reply:
Would it help if I temporarily decreased the refresh token lifetime for your app client so that you could do some accelerated testing?

Reply:
sure. If you can set it per credential then set the hermes-dev one to 1 day, or even just 1 hour would be fine.

Reply:
I've made the following changes to your app client:

<img width="558" alt="Screenshot 2023-11-09 at 08 02 11" src="https://github.com/LCOGT/hermes/assets/728407/a83218c9-2b0d-4e34-ac9a-961f27f58a4f">

Reply:
Thanks for that, I can confirm the code works as intended, removing the token for the user when its been over the 60 minutes, forcing the user to re-auth on their profile page

Reply:
I restored it to the default settings:

![Screenshot 2023-11-10 at 08 42 53](https://github.com/LCOGT/hermes/assets/728407/6e35cde4-cf26-4b26-b8e8-9c31bd3967eb)